src/main/java/org/openjdk/jextract/impl/CDeclarationPrinter.java

```diff
@@ -0,0 +1,261 @@
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package org.openjdk.jextract.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.openjdk.jextract.Declaration;
+import org.openjdk.jextract.Type;
+
+final class CDeclarationPrinter implements Declaration.Visitor<Void, Void> {
+    private static String SPACES = " ".repeat(92);
+    private int align = 0;
+    private String prefix;
+
+    private void incr() {
+        align += 4;
+    }
+
+    private void decr() {
+        align -= 4;
+    }
+
+    private CDeclarationPrinter(String prefix) {
+        this.prefix = prefix;
+    }
+
+    private void indent() {
+        builder.append(prefix);
+        builder.append(SPACES.substring(0, align));
+    }
+
+    private final StringBuilder builder = new StringBuilder();
+
+    private String print(Declaration decl) {
+        decl.accept(this, null);
+        return builder.toString();
+    }
+
+    // Return C source style signature for the given declaration.
+    // The prefix is emitted for every line. This can be used
+    // to prefix per line comment character "*" in generated javadoc.
+    static String declaration(Declaration decl, String prefix) {
+        Objects.requireNonNull(decl);
+        Objects.requireNonNull(prefix);
+        return new CDeclarationPrinter(prefix).print(decl);
+    }
+
+    static String declaration(Type.Function funcType, String name) {
+        return nameAndType(funcType, "*" + name);
+    }
+
+    @Override
+    public Void visitScoped(Declaration.Scoped d, Void ignored) {
+        indent();
+        var tag = typeTag(d);
+        if (!tag.isEmpty()) {
+            builder.append(tag);
+            if (!d.name().isEmpty()) {
+                builder.append(" " + d.name());
+            }
+            builder.append(" {");
+            builder.append("\n");
+            incr();
+        }
+        d.members().forEach(m -> m.accept(this, null));
+        if (!tag.isEmpty()) {
+            decr();
+            indent();
+            builder.append("};\n");
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitFunction(Declaration.Function d, Void ignored) {
+        indent();
+        var ftype = d.type();
+        var rtype = ftype.returnType();
+        builder.append(nameAndType(rtype, ""));
```

Comment:
I would rather add another helper function to just print a type, rather than reusing `nameAndType` here. Also, we should check that the type doesn't have a name attached.